File: IcarianCS/src/Random.cs
// Icarian Engine - C# Game Engine
// 
// License at end of file.

using IcarianEngine.Maths;
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

namespace IcarianEngine
{
    public static class Random
    {
        [MethodImpl(MethodImplOptions.InternalCall)]
        extern static void FillBuffer(byte[] a_buffer, uint a_offset);

        const uint BufferSize = 512;

        static byte[] s_buffer = new byte[BufferSize];
        static uint s_bufferIndex = BufferSize;

        static void FillBuffer()
        {
            uint offset = BufferSize - s_bufferIndex;
            for (int i = 0; i < offset; ++i)
            {
                s_buffer[i] = s_buffer[i + s_bufferIndex];
            }

            FillBuffer(s_buffer, offset);

            s_bufferIndex = 0;
        }

        /// <summary>
        /// Generates a random int
        /// </summary>
        /// <returns>A random int value</returns>
        public static int Int()
        {
            if (s_bufferIndex + 4 >= BufferSize)
            {
                FillBuffer();
            }

            int value = BitConverter.ToInt32(s_buffer, (int)s_bufferIndex);

            s_bufferIndex += 4;

            return value;
        }
        /// <summary>
        /// Generates a random uint
        /// </summary>
        /// <returns>A random uint value</returns>
        public static uint UInt()
        {
            if (s_bufferIndex + 4 >= BufferSize)
            {
                FillBuffer();
            }

            uint value = BitConverter.ToUInt32(s_buffer, (int)s_bufferIndex);

            s_bufferIndex += 4;

            return value;
        }

        /// <summary>
        /// Generates a random uint in range
        /// </summary>
        /// <param name="a_min">The minimum value inclusive</param>
        /// <param name="a_max">The maximum value inclusive</param>
        /// <returns>A random value in the range</returns>
        public static uint Range(uint a_min, uint a_max)
        {
            return a_min + (UInt() % (a_max - a_min + 1));
        }
        /// <summary>
        /// Generates a random float in range
        /// </summary>
        /// <param name="a_min">The minimum value inclusive</param>
        /// <param name="a_max">The maximum value inclusive</param>
        /// <returns>A random value in the range</returns>
        public static float Range(float a_min, float a_max)
        {
            double mul = UInt() / (double)uint.MaxValue;
            float range = a_max - a_min;

            return (float)(a_min + (mul * range));
        }
        /// <summary>
        /// Generates a random int in range
        /// </summary>
        /// <param name="a_min">The minimum value inclusive</param>
        /// <param name="a_max">The maximum value inclusive</param>
        /// <returns>A random value in the range</returns>
        public static int Range(int a_min, int a_max)
        {
            return a_min + (Int() % (a_max - a_min + 1));
        }

        /// <summary>
        /// Generates a random <see cref="IcarianEngine.Maths.Vector2" /> in range
        /// </summary>
        /// <param name="a_min">The minimum value inclusive</param>
        /// <param name="a_max">The maximum value inclusive</param>
        /// <returns>A random value in the range</returns>
        public static Vector2 Range(Vector2 a_min, Vector2 a_max)
        {
            return new Vector2(Range(a_min.X, a_max.X), Range(a_min.Y, a_max.Y));
        }
        /// <summary>
        /// Generates a random <see cref="IcarianEngine.Maths.Vector3" /> in range
        /// </summary>
        /// <param name="a_min">The minimum value inclusive</param>
        /// <param name="a_max">The maximum value inclusive</param>
        /// <returns>A random value in the range</returns>
        public static Vector3 Range(Vector3 a_min, Vector3 a_max)
        {
            return new Vector3(Range(a_min.X, a_max.X), Range(a_min.Y, a_max.Y), Range(a_min.Z, a_max.Z));
        }
        /// <summary>
        /// Generates a random <see cref="IcarianEngine.Maths.Vector4" /> in range
        /// </summary>
        /// <param name="a_min">The minimum value inclusive</param>
        /// <param name="a_max">The maximum value inclusive</param>
        /// <returns>A random value in the range</returns>
        public static Vector4 Range(Vector4 a_min, Vector4 a_max)
        {
            return new Vector4(Range(a_min.X, a_max.X), Range(a_min.Y, a_max.Y), Range(a_min.Z, a_max.Z), Range(a_min.W, a_max.W));
        }

        /// <summary>
        /// Generates a random <see cref="IcarianEngine.Maths.Quaternion" />
        /// </summary>
        /// <returns>A normalized random <see cref="IcarianEngine.Maths.Quaternion" /></returns>
        public static Quaternion Rotation()
        {
            return Quaternion.Normalized(new Quaternion(Range(-1.0f, 1.0f), Range(-1.0f, 1.0f), Range(-1.0f, 1.0f), Range(-1.0f, 1.0f)));
        }

        /// <summary>
        /// Generates a random <see cref="IcarianEngine.Maths.IVector2" /> in range
        /// </summary>
        /// <param name="a_min">The minimum value inclusive</param>
        /// <param name="a_max">The maximum value inclusive</param>
        /// <returns>A random value in the range</returns>
        public static IVector2 Range(IVector2 a_min, IVector2 a_max)
        {
            return new IVector2(Range(a_min.X, a_max.X), Range(a_min.Y, a_max.Y));
        }
        /// <summary>
        /// Generates a random <see cref="IcarianEngine.Maths.IVector3" /> in range
        /// </summary>
        /// <param name="a_min">The minimum value inclusive</param>
        /// <param name="a_max">The maximum value inclusive</param>
        /// <returns>A random value in the range</returns>
        public static IVector3 Range(IVector3 a_min, IVector3 a_max)
        {
            return new IVector3(Range(a_min.X, a_max.X), Range(a_min.Y, a_max.Y), Range(a_min.Z, a_max.Z));
        }
        /// <summary>
        /// Generates a random <see cref="IcarianEngine.Maths.IVector4" /> in range
        /// </summary>
        /// <param name="a_min">The minimum value inclusive</param>
        /// <param name="a_max">The maximum value inclusive</param>
        /// <returns>A random value in the range</returns>
        public static IVector4 Range(IVector4 a_min, IVector4 a_max)
        {
            return new IVector4(Range(a_min.X, a_max.X), Range(a_min.Y, a_max.Y), Range(a_min.Z, a_max.Z), Range(a_min.W, a_max.W));
        }
    }
}

// MIT License
// 
// Copyright (c) 2024 River Govers
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

File: IcarianNative/src/Random.cpp
// Icarian Engine - C# Game Engine
// 
// License at end of file.

#include "Random.h"

#include "IcarianError.h"
#include "Runtime/RuntimeManager.h"

static Random* Instance = nullptr;

RUNTIME_FUNCTION(void, Random, FillBuffer,
{
    const uint32_t size = mono_array_length(a_buffer);
    const uint32_t fillSize = size - a_offset;

    const uint8_t* bytes = Random::GetBytes(fillSize);

    for (uint32_t i = 0; i < fillSize; ++i)
    {
        mono_array_set(a_buffer, mono_byte, i + a_offset, bytes[i]);
    }
}, MonoArray* a_buffer, uint32_t a_offset)

Random::Random()
{
    m_index = BufferSize;
    
    FillBuffer();
}
Random::~Random()
{
    
}

void Random::Init()
{
    if (Instance == nullptr)
    {
        Instance = new Random();

        BIND_FUNCTION(IcarianEngine, Random, FillBuffer);
    }
}
void Random::Destroy()
{
    if (Instance != nullptr)
    {
        delete Instance;
        Instance = nullptr;
    }
}

void Random::FillBuffer()
{
    const uint32_t offset = BufferSize - m_index;

    for (uint32_t i = 0; i < offset; ++i)
    {
        m_buffer[i] = m_buffer[m_index + i];
    }

    const uint32_t size = BufferSize - offset;

    // TODO: Use better RNG as rand seems to be bias 
    // rand does not seem to be very good so have to do it at the byte level to increase the noise
    uint8_t* writeBuffer = (uint8_t*)(m_buffer + offset);
    for (uint32_t i = 0; i < size; ++i)
    {
        writeBuffer[i] = (uint8_t)((double)rand() / RAND_MAX * UINT8_MAX);
    }

    m_index = 0;
}

uint8_t* Random::GetBytes(uint32_t a_size)
{
    IVERIFY(a_size < BufferSize);

    if (Instance->m_index + a_size > BufferSize)
    {
        Instance->FillBuffer();
    }

    uint8_t* buffer = Instance->m_buffer + Instance->m_index;
    Instance->m_index += a_size;

    return buffer;
}

uint32_t Random::Range(uint32_t a_min, uint32_t a_max)
{
    if (Instance->m_index + sizeof(uint32_t) >= BufferSize)
    {
        Instance->FillBuffer();
    }

    uint32_t* buffer = (uint32_t*)(Instance->m_buffer + Instance->m_index);
    Instance->m_index += sizeof(uint32_t);

    return a_min + (*buffer % (a_max - a_min));
}
float Random::Range(float a_min, float a_max)
{
    if (Instance->m_index + sizeof(float) >= BufferSize)
    {
        Instance->FillBuffer();
    }

    uint32_t* buffer = (uint32_t*)(Instance->m_buffer + Instance->m_index);
    Instance->m_index += sizeof(uint32_t);

    return (float)(a_min + ((double)*buffer / UINT32_MAX) * (a_max - a_min));
}
int32_t Random::Range(int32_t a_min, int32_t a_max)
{
    if (Instance->m_index + sizeof(int32_t) >= BufferSize)
    {
        Instance->FillBuffer();
    }

    int32_t* buffer = (int32_t*)(Instance->m_buffer + Instance->m_index);
    Instance->m_index += sizeof(int32_t);

    return a_min + (*buffer % (a_max - a_min));
}

// MIT License
// 
// Copyright (c) 2024 River Govers
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.